Assignments/ip2/ip2.md

```diff
@@ -0,0 +1,310 @@
+
+## Implementation Tasks
+This deliverable has four parts; each part will be graded on its own rubric. You should complete the assignment one part at a time, in the order presented here.
+
+**General Requirements**: Implement your code *only* in the files specified:
+* Task 1: `frontend/src/classes/interactable/TicTacToeAreaController.ts`
+* Task 1: `frontend/src/classes/interactable/TicTacToeAreaController.test.ts`
+* Task 2: `frontend/src/components/Town/interactables/TicTacToe/TicTacToeArea.tsx`
+* Task 3: `frontend/src/components/Town/interactables/TicTacToe/TicTacToeBoard.tsx`
+* Task 4: `frontend/src/components/Town/interactables/Leaderboard.tsx`
+
+You should not install any additional dependencies. The autograder will ignore any other files that you modify, and will not install any dependencies that you add to the project.
+
+### Task 1: The TicTacToeAreaController (24 points)
+The `TicTacToeAreaController` is a class that is responsible for managing the state of a single TicTacToe game. It is responsible for communicating with the TownService backend. Frontend components will interact with the `TicTacToeAreaController` to get the current state of the game, and to send commands to the backend to update the game state. The `TicTacToeAreaController` also will emit events when the state of the game changes, so that frontend components can update their state accordingly.
+
+`TicTacToeAreaController` extends the base `GameAreaController` class.
+The base class tracks the game model (`this._model`), the set of players in the game (`this.players`), and the set of observers in the game (`this.observers`). It also provides methods for joining and leaving the game, as well as a base implementation of `_updateFrom`, which is responsible for updating the game state when the backend notifies the frontend of a change.
+
+Your first task is to implement each of the properties and methods of the `TicTacToeAreaController` class (`frontend/src/classes/interactable/TicTacToeAreaController.ts`). The specification for these properties and methods appears below:
+
+{::options parse_block_html="true" /}
+<details><summary markdown="span">View the specification for these methods</summary>
+{% highlight typescript %}
+  {
+  /**
+   * Returns the current state of the board.
+   *
+   * The board is a 3x3 array of TicTacToeCell, which is either 'X', 'O', or undefined.
+   *
+   * The 2-dimensional array is indexed by row and then column, so board[0][0] is the top-left cell,
+   * and board[2][2] is the bottom-right cell
+   */
+  get board(): TicTacToeCell[][]
+  /**
+   * Returns the player with the 'X' game piece, if there is one
+   */
+  get x(): PlayerController | undefined
+  /**
+   * Returns the player with the 'O' game piece, if there is one
+   */
+  get o(): PlayerController | undefined 
+  /**
+   * Returns the number of moves that have been made in the game
```

Comment:
_model.game can be undefined. Should we specify a default return value (0) if that is the case for additional clarity?

Reply:
Similar case with `get status(): GameStatus` as well. The tests expect `get status()` to return `WAITING_TO_START`. I think we should clarify this in the doc.